src/chains/chains.controller.spec.ts

```diff
@@ -1,22 +1,60 @@
 import { Test, TestingModule } from '@nestjs/testing';
-import { ConfigServiceModule } from '../services/config-service/config-service.module';
 import { ChainsController } from './chains.controller';
 import { ChainsService } from './chains.service';
+import { Backbone, Chain, Page } from './entities';
 
-describe('ChainsController', () => {
-  let controller: ChainsController;
+describe('ChainsController (Unit)', () => {
+  let chainsController: ChainsController;
+  let chainsService: ChainsService;
+
+  const chainsResponse: Page<Chain> = {
+    count: 2,
+    next: null,
+    previous: null,
+    results: [
+      <Chain>{
+        chainId: '1',
+        chainName: 'testChain',
+        vpcTransactionService: 'http://test-endpoint.local',
+      },
+    ],
+  };
+
+  const backboneResponse: Backbone = {
+    name: 'Service Name',
+    version: '4.6.1',
+    api_version: 'v1',
+    secure: false,
+    host: 'service.host',
+    headers: ['header1', 'header2'],
+    settings: { key1: 'value1', key2: 'value2' },
+  };
 
   beforeEach(async () => {
     const module: TestingModule = await Test.createTestingModule({
-      imports: [ConfigServiceModule],
       controllers: [ChainsController],
-      providers: [ChainsService],
+      providers: [
+        {
+          provide: ChainsService,
+          useValue: {
+            getChains: jest.fn().mockResolvedValue(chainsResponse),
+            getBackbone: jest.fn().mockResolvedValue(backboneResponse),
+          },
```

Comment:
Here we can use the real `ChainsService`

You'd still need to mock `axios` unless you want to isolate the `ChainsController`.

Reply:
@fmrsabino 
I was thinking in these tests as _controller tests_ in isolation, but later I reviewed what you did for `balances.controller.spec.ts` and maybe makes sense to have this `*.controller.spec.ts` as some kind of _almost-e2e-tests_ where nothing is mocked except the external service's data provider in the most outer layer.

I think it's ok by now, but I have to admit that I have some concerns about how this will evolve in time. I mean, we're coding something like the external service will be called one time with this arguments, later it will be called with these other arguments, later another call which would return this thing, etc. etc. Maybe it's too tied to the concrete implementation we have now, and if for any reason the service logic changes these tests would be hard to maintain. 

Plus, if a payload coming from an external service changes, our tests will be passing anyway.

I'm wondering if having isolated unit tests _per layer_, and real _e2e tests_ (where nothing is mocked, even the external services) would be more maintainable. With the real e2e tests you only have to code the expected result the client would receive, and coupling between tests and implementation is near 0.

(Sorry for the long comment) 😅

Reply:
> @fmrsabino I was thinking in these tests as _controller tests_ in isolation, but later I reviewed what you did for `balances.controller.spec.ts` and maybe makes sense to have this `*.controller.spec.ts` as some kind of _almost-e2e-tests_ where nothing is mocked except the external service's data provider in the most outer layer.
> 
> I think it's ok by now, but I have to admit that I have some concerns about how this will evolve in time. I mean, we're coding something like the external service will be called one time with this arguments, later it will be called with these other arguments, later another call which would return this thing, etc. etc. Maybe it's too tied to the concrete implementation we have now, and if for any reason the service logic changes these tests would be hard to maintain.
> 
> Plus, if a payload coming from an external service changes, our tests will be passing anyway.

The idea was to Unit test the the endpoint without having any external dependencies. The mocks on the HTTP level are required just because we need some "fake" http service where we say that if X is requested return Y. Besides that there's nothing that ties to the implementation.

If we isolate the test (each Unit test) then it is more tied to the implementation because we need to mock each dependency and if the implementation changes we also need to change the test.

I'm aware that mocking the Axios client can be verbose and we can think of ways to improve it but I think these tests are valuable in a way that they treat the actual implementation as a black box where we just need to write the endpoint that we are hitting, the responses of the external services would be and the expected response. If change the implementation the test should still pass (unless we change how the external services usage).

> I'm wondering if having isolated unit tests _per layer_, and real _e2e tests_ (where nothing is mocked, even the external services) would be more maintainable. With the real e2e tests you only have to code the expected result the client would receive, and coupling between tests and implementation is near 0.

My approach with the tests in `*.controller.spec.ts` does not mean that we should not have tests that test components in isolation. I find those tests still really valuable 👍 

For real E2E tests it can be more complex... specially when creating transactions and submitting transactions. We need to have some kind of test database or similar infrastructure to run this. Even though they provide a lot of value, some flows can be tricky to test.

Reply:
> If we isolate the test (each Unit test) then it is more tied to the implementation because we need to mock each dependency and if the implementation changes we also need to change the test.

I partially agree, I think too much spying would led to coupling between tests and implementation, but by just mocking the dependency you're just tied to its interface, not its internal implementation. By mocking axios, which is an internal detail of the dependency (in the future the dependency could even use a different http client) you're tied to the actual implementation of the dependency. And the test have to know how an internal detail of the dependency needs to operate (which calls needs to make, to where, with which arguments, which should expect in the response, etc.)

Just to clear a bit my view 🙂: for me the http services we're calling (config service and transaction service) are just data providers, they just provide data, the protocol (http) it's just a detail, and could be treated like a database, a queue, etc. In my mind mocking `axios` here is the same that mocking `mongodb` driver for example.

But I get your point, I see these tests valuable as well, and maybe when e2e test are in place we'll see we're doing something very close to this, and we can refactor. I will implement the test as you did for `balances.controller.spec.ts` 👍 

> For real E2E tests it can be more complex... specially when creating transactions and submitting transactions. We need to have some kind of test database or similar infrastructure to run this. Even though they provide a lot of value, some flows can be tricky to test.

Got it, OK, you have all the experience on those scenarios. As I said my main concern was we were doing an hybrid approach between unit and e2e tests, and when we have real e2e tests all this code would be duplicated. But we'll tackle that in the future. 

Thanks for discussing this, glad to share opinions 😄

Reply:
> I partially agree, I think too much spying would led to coupling between tests and implementation, but by just mocking the dependency you're just tied to its interface, not its internal implementation. By mocking axios, which is an internal detail of the dependency (in the future the dependency could even use a different http client) you're tied to the actual implementation of the dependency. And the test have to know how an internal detail of the dependency needs to operate (which calls needs to make, to where, with which arguments, which should expect in the response, etc.)
> Just to clear a bit my view 🙂: for me the http services we're calling (config service and transaction service) are just data providers, they just provide data, the protocol (http) it's just a detail, and could be treated like a database, a queue, etc. In my mind mocking `axios` here is the same that mocking `mongodb` driver for example.
> 
> But I get your point, I see these tests valuable as well, and maybe when e2e test are in place we'll see we're doing something very close to this, and we can refactor. I will implement the test as you did for `balances.controller.spec.ts` 👍

Ah completely agree! Mocking `axios` does seem weird. My original idea was to have a `TestHttpModule` which is used instead of the actual one. Or create an interface of our own. This is something that we should sync on.

The same would be the same for external "services" like caching solutions or even the environment that we use (`.env`).

For testing it should be easy enough to replace these external components with a test one but one that is not tied directly to the "actual" implementation (in this case mocking `axios`).

> > For real E2E tests it can be more complex... specially when creating transactions and submitting transactions. We need to have some kind of test database or similar infrastructure to run this. Even though they provide a lot of value, some flows can be tricky to test.
> 
> Got it, OK, you have all the experience on those scenarios. As I said my main concern was we were doing an hybrid approach between unit and e2e tests, and when we have real e2e tests all this code would be duplicated. But we'll tackle that in the future.
> 
> Thanks for discussing this, glad to share opinions 😄

I'd still consider the tests that I mentioned above as "unit" tests (the unit is just bigger instead of being a single class) but there's no golden rule for this. We should get to a solution that it's easy to read and maintain 🙌